File: extensions/syncforreddit/src/main/java/app/revanced/extension/syncforreddit/FixSLinksPatch.java
package app.revanced.extension.syncforreddit;

import com.laurencedawson.reddit_sync.ui.activities.WebViewActivity;

import app.revanced.extension.shared.fixes.slink.BaseFixSLinksPatch;

/** @noinspection unused*/
public class FixSLinksPatch extends BaseFixSLinksPatch {
    static {
        INSTANCE = new FixSLinksPatch();
    }

    private FixSLinksPatch() {
        webViewActivityClass = WebViewActivity.class;
    }

    public static boolean patchResolveSLink(String link) {
        return INSTANCE.resolveSLink(link);
    }

    public static void patchSetAccessToken(String accessToken) {
        INSTANCE.setAccessToken(accessToken);
    }
}

File: src/main/kotlin/app/revanced/patches/reddit/customclients/sync/syncforreddit/fix/slink/FixSLinksPatch.kt
package app.revanced.patches.reddit.customclients.sync.syncforreddit.fix.slink

import app.revanced.patcher.extensions.InstructionExtensions.addInstructions
import app.revanced.patcher.patch.bytecodePatch
import app.revanced.patches.reddit.customclients.sync.syncforreddit.fix.slink.fingerprints.linkHelperOpenLinkFingerprint

@Suppress("unused")
val fixSLinksPatch = bytecodePatch(
    name = "Fix /s/ links",
    description = "Fixes the issue where /s/ links do not work.",
    requiresIntegrations = true,
) {
    compatibleWith(
        "com.laurencedawson.reddit_sync",
        "com.laurencedawson.reddit_sync.pro",
        "com.laurencedawson.reddit_sync.dev",
    )

    val linkHelperOpenLinkResult by linkHelperOpenLinkFingerprint

    execute {
        linkHelperOpenLinkResult.mutableMethod.addInstructions(
            1,
            """
                invoke-static { p3 }, Lapp/revanced/integrations/syncforreddit/FixSLinksPatch;->resolveSLink(Ljava/lang/String;)Ljava/lang/String;
                move-result-object p3
            """,
        )
    }
}
